superset/models/slice.py

```diff
@@ -247,6 +247,17 @@ def form_data(self) -> Dict[str, Any]:
         update_time_range(form_data)
         return form_data
 
+    def get_query_context(self) -> Dict[str, Any]:
+        query_context: Dict[str, Any] = {}
+        if not self.query_context:
+            return query_context
+        try:
+            query_context = json.loads(self.query_context)
+        except json.decoder.JSONDecodeError as ex:
+            logger.error("Malformed json in slice's query context", exc_info=True)
+            logger.exception(ex)
+        return query_context
+
```

Comment:
We can make this function more generic. like this:

```Python
    def get_query_context(self) -> QueryContext:
        if self.query_context:
            try:
                return self.query_context(**json.loads(self.query_context))
            except json.decoder.JSONDecodeError as ex:
                logger.error("Malformed json in slice's query context", exc_info=True)
                logger.exception(ex)
        return QueryContext(
                       datasource={"type": self.datasource_type, "id": self.datasource_id},
                       queries=[self.viz.query_obj()],)
```

Reply:
@zhaoyongjie on the first return statement I believe we could instantiate the `QueryContext`. Also, for non v1 charts or in the case there isn't a query context payload, we probably should just return `None`, as the legacy charts aren't really compatible with `QueryContext`. Could this work?
```python
    def get_query_context(self) -> Optional[QueryContext]:
        if self.query_context:
            try:
                return QueryContext(**json.loads(self.query_context))
            except json.decoder.JSONDecodeError as ex:
                logger.error("Malformed json in slice's query context", exc_info=True)
                logger.exception(ex)
        return None
```

Reply:
Nice! It makes sense return `None` in legacy chart.

Reply:
Thanks for suggestions! I made the change, would appreciate another look 🙂